File: Go/MaximumGap.go
//go:build ignore

package main

import "fmt"

func maximumGap(nums []int) int {
	n := len(nums)
	if n < 2 {
		return 0
	}
	mn, mx := nums[0], nums[0]
	for _, x := range nums {
		if x < mn {
			mn = x
		}
		if x > mx {
			mx = x
		}
	}
	if mn == mx {
		return 0
	}

	// Pigeonhole: max gap >= ceil((mx-mn)/(n-1)), so it must occur between buckets of that size.
	bucketSize := max((mx-mn)/(n-1), 1)
	numBuckets := (mx-mn)/bucketSize + 1
	bMin := make([]int, numBuckets)
	bMax := make([]int, numBuckets)
	seen := make([]bool, numBuckets)

	for _, x := range nums {
		i := (x - mn) / bucketSize
		if !seen[i] {
			bMin[i], bMax[i] = x, x
			seen[i] = true
		} else {
			if x < bMin[i] {
				bMin[i] = x
			}
			if x > bMax[i] {
				bMax[i] = x
			}
		}
	}

	best := 0
	prevMax := mn
	for i := range numBuckets {
		if !seen[i] {
			continue
		}
		if gap := bMin[i] - prevMax; gap > best {
			best = gap
		}
		prevMax = bMax[i]
	}
	return best
}

func main() {
	fmt.Println(maximumGap([]int{3, 6, 9, 1}))  // 3
	fmt.Println(maximumGap([]int{10}))          // 0
	fmt.Println(maximumGap([]int{1, 1, 1, 1}))  // 0
	fmt.Println(maximumGap([]int{1, 10000000})) // 9999999
}

File: Java/MaximumGap.java
public class MaximumGap {
    public int maximumGap(int[] nums) {
        int n = nums.length;
        if (n < 2) return 0;
        int mn = nums[0], mx = nums[0];
        for (int x : nums) {
            if (x < mn) mn = x;
            if (x > mx) mx = x;
        }
        if (mn == mx) return 0;

        // Pigeonhole: max gap >= ceil((mx-mn)/(n-1)), so it lies between buckets of that size.
        int bucketSize = Math.max(1, (mx - mn) / (n - 1));
        int numBuckets = (mx - mn) / bucketSize + 1;
        int[] bMin = new int[numBuckets];
        int[] bMax = new int[numBuckets];
        boolean[] seen = new boolean[numBuckets];

        for (int x : nums) {
            int i = (x - mn) / bucketSize;
            if (!seen[i]) {
                bMin[i] = x;
                bMax[i] = x;
                seen[i] = true;
            } else {
                bMin[i] = Math.min(bMin[i], x);
                bMax[i] = Math.max(bMax[i], x);
            }
        }

        int best = 0, prevMax = mn;
        for (int i = 0; i < numBuckets; i++) {
            if (!seen[i]) continue;
            best = Math.max(best, bMin[i] - prevMax);
            prevMax = bMax[i];
        }
        return best;
    }

    public static void main(String[] args) {
        MaximumGap s = new MaximumGap();
        System.out.println(s.maximumGap(new int[] { 3, 6, 9, 1 }));  // 3
        System.out.println(s.maximumGap(new int[] { 10 }));          // 0
        System.out.println(s.maximumGap(new int[] { 1, 1, 1, 1 }));  // 0
        System.out.println(s.maximumGap(new int[] { 1, 10000000 })); // 9999999
    }
}
